src/kernel/src/machine/arm/virt/processor.rs

```diff
@@ -0,0 +1,68 @@
+use core::str::FromStr;
+
+use arm64::registers::MPIDR_EL1;
+use registers::interfaces::Readable;
+
+use crate::machine::info::devicetree;
+
+#[derive(Debug, Default, Copy, Clone, PartialEq)]
+pub enum BootMethod {
+    Psci,
+    SpinTable,
+    #[default]
+    Unknown,
+}
+
+impl BootMethod {
+    fn as_str(&self) -> &'static str {
+        match self {
+            Self::Psci => "psci",
+            Self::SpinTable => "spintable",
+            Self::Unknown => "unknown",
+        }
+    }
+}
+
+impl FromStr for BootMethod {
+    type Err = ();
+
+    // Required method
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match s {
+            "psci" => Ok(BootMethod::Psci),
+            "spin-table" => Ok(BootMethod::SpinTable),
+            _ => Err(())
+        }
+    }
+}
+
+pub fn enumerate_cpus() -> u32 {
```

Comment:
Is the return a bit field of 32 bits?  If so we might want to make this a much larger value.  On Arm this might not be > 32 cores, but on Intel...

Reply:
No, it is a scalar value. So we can count way more than 32 cores.